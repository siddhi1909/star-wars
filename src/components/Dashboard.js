// IMPORT EXTERNAL LIBRARIES/MODULES
import React, {useState, useEffect, useContext, useRef} from 'react';
import {useHistory} from "react-router-dom";
import {ReactReduxContext, useDispatch} from 'react-redux';
// IMPORT API & ROUTE ACTIONS
import {usePlanetSearch, usePlanetDetails} from '../effects';
import {loginUserFailure} from '../redux/actions/authActions';
import Loader from './Loader'
import PlanetDetail from './PlanetDetail'
import Toast from './Toast'
import Pagination from "./Pagination";
import PlanetList from "./PlanetList";
import SimpleCrypto from "simple-crypto-js";

const Dashboard = () => {
    const {store} = useContext(ReactReduxContext);
    let _secretKey = "siddhi";
    let simpleCrypto = new SimpleCrypto(_secretKey);
    const userName = store.getState().auth.user || (localStorage.getItem('user')?simpleCrypto.decrypt(localStorage.getItem('user')):'');
    const dispatch = useDispatch();
    const [{results, isLoading, count, error}, searchPlanet] = usePlanetSearch();
    const [{planet, isPlanetLoading}, getPlanet] = usePlanetDetails();
    const [searchText, setSearchText] = useState('');
    const [currentPage, setCurrentPage] = useState(1);
    const [recordsPerPage] = useState(10);
    const history = useHistory();
    let planetList = [...results] || {};
    const ref = useRef(null);
    const [width, setWidth] = useState(0);
    const [showToast, setShowToast] = useState(false);
    const [messageToast, setMessageToast] = useState('');
    const [levelToast, setLevelToast] = useState('');
    const [searchCount, setSearchCount] = useState(0);

    useEffect(() => {
        searchPlanet('', 1);
        let widthCalc = ref.current ? ref.current.offsetWidth : 0;
        setWidth(widthCalc);
    }, [ref.current]);

    useEffect(() => {
        if (results && results.length <= 0 && error !== '') {
            showToastFunction(error, 'warning');
        }
    }, [results, error]);

    useEffect(() => {
        if (searchCount > 15) {
            showToastFunction('Sorry! You cannot make more than 15 searched in a minute.', 'warning');
        } else if (showToast) {
            setShowToast(false);
        }
    }, [searchCount]);

    useEffect(() => {
        if (userName !== 'Luke Skywalker') {
            const interval = setInterval(() => {
                setSearchCount(0);
            }, 3000);
            return () => clearInterval(interval);
        }
    }, [userName]);

    /*
    * onLogout function for user Loagout
    * */
    const onLogout = () => {
        /* removes user from Localstorage */
        localStorage.removeItem('user');
        dispatch(loginUserFailure());
        history.push("/");
    }

    /*
    * Handles search of Planets
    * @param {e} event Event on input
    * @return nothing
    * */
    const handleSearch = (e) => {
        setCurrentPage(1);
        setSearchText(e.target.value);
        if ((e.target.value.length > 2 || e.target.value.length === 0) &&
            ((userName !== 'Luke Skywalker' && searchCount <= 15) || (userName === 'Luke Skywalker'))
        ) {
            searchPlanet(e.target.value, 1);
            if (userName !== 'Luke Skywalker') {
                setSearchCount(searchCount + 1);
            }
        }
    }

    /*
    * @param {message}  message to show on Toast
    * @param {level}  level of message like success, warning or error
    * */
    const showToastFunction = (message, level) => {
        setLevelToast(level);
        setMessageToast(message);
        setShowToast(true);
    }
    // useEffect(() => {
    //     setTimeout(() => {
    //         setShowToast(false);
    //     }, 3000)
    // }, [showToast])

    return (
        <>
            <div className="container-fluid sticky-top navbar-light bg-light padding-top-bottom-10">
                <div className="row">
                    <div className="col-md-3 col-sm-1"></div>
                    <div className="col-md-6 col-sm-8">
                        <input type="text" className="form-control" id="inputSearch"
                               placeholder="Search planets!"
                               value={searchText} onChange={handleSearch}/>
                    </div>
                    <div className="col-md-3 col-sm-3 text-right">
                        <span className="user-dashboard">Hi! {userName}</span>
                        <button onClick={onLogout} className="btn btn-outline-success float-right btn-sm"
                                type="button">Logout
                        </button>
                    </div>
                </div>
            </div>
            <div className="container padding-10-0">
                <div className="row">
                    <div className="col-md-7 text-center" ref={ref}>
                        {isLoading ? (
                                <Loader/>
                            ) :
                            (<div className="hide-overflow planet-list-outer">
                                <PlanetList planetList={planetList} getPlanet={getPlanet} width={width}/>
                                <Pagination count={count} recordsPerPage={recordsPerPage} currentPage={currentPage} planetList={planetList} searchText={searchText} setCurrentPage={setCurrentPage} searchPlanet={searchPlanet}/>
                            </div>)
                        }
                    </div>
                    <div className="col-md-5 outer-div text-center">
                        {isPlanetLoading ?
                            (<Loader/>) :
                            <PlanetDetail planet={planet} />
                        }
                    </div>
                </div>
            </div>
            <Toast
                level={levelToast}
                message={messageToast}
                visible={showToast}
            />
        </>
    );
}

export default Dashboard;