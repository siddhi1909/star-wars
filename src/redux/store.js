// // IMPORT EXTERNAL LIBRARIES/MODULES
import {createStore} from 'redux';
// IMPORT ROOT REDUCER
import rootReducer from './reducer/rootReducer';

const initial_state = {};
// create the store
const store = createStore(rootReducer, initial_state);

export default store;
