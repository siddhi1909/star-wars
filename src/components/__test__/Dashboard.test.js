// IMPORT EXTERNAL LIBRARIES/MODULES
import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from "react-redux";
import {BrowserRouter as Router} from "react-router-dom";
import { shallow, configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
// IMPORT COMPONENTS, API & ROUTE ACTIONS
import Dashboard from '../Dashboard';
import store from "../../redux/store";

configure({ adapter: new Adapter() });

describe('>>>D A S H B O A R D --- Render REACT COMPONENTS',()=>{
    let wrapper

    beforeEach(()=>{
        wrapper = mount(<Provider store={store}><Router><Dashboard/></Router></Provider>)
    })

    it('+++ render the App component', () => {
        expect(wrapper.find('Dashboard').length).toEqual(1)
    });

    it('+++ render < Dashboard /> renders correctly', () => {
        const renderedValue =  renderer.create(<Provider store={store}><Router><Dashboard/></Router></Provider>).toJSON()
        expect(renderedValue).toMatchSnapshot();
    });

});