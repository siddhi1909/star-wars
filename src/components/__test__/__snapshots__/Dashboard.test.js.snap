// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>D A S H B O A R D --- Render REACT COMPONENTS +++ render < Dashboard /> renders correctly 1`] = `
Array [
  <div
    className="container-fluid sticky-top navbar-light bg-light padding-top-bottom-10"
  >
    <div
      className="row"
    >
      <div
        className="col-md-3 col-sm-1"
      />
      <div
        className="col-md-6 col-sm-8"
      >
        <input
          className="form-control"
          id="inputSearch"
          onChange={[Function]}
          placeholder="Search planets!"
          type="text"
          value=""
        />
      </div>
      <div
        className="col-md-3 col-sm-3 text-right"
      >
        <span
          className="user-dashboard"
        >
          Hi! 
          
        </span>
        <button
          className="btn btn-outline-success float-right btn-sm"
          onClick={[Function]}
          type="button"
        >
          Logout
        </button>
      </div>
    </div>
  </div>,
  <div
    className="container padding-10-0"
  >
    <div
      className="row"
    >
      <div
        className="col-md-7 text-center"
      >
        <div
          className="hide-overflow planet-list-outer"
        >
          <div>
            No planets
          </div>
          
        </div>
      </div>
      <div
        className="col-md-5 outer-div text-center"
      >
        
      </div>
    </div>
  </div>,
  <div
    className="toast  "
  >
    <figure>
      <img
        alt=""
        src="http://svgshare.com/i/19y.svg"
      />
    </figure>
    <p>
      
    </p>
  </div>,
]
`;
